BlueBrain/Search
# Luigi pipeline sketch

Fixes #562 and fixes #558

## Description
### Diagrams
Folder naming logic
![Screenshot 2022-02-21 at 14 04 51](https://user-images.githubusercontent.com/18519371/154960967-3335acd9-d8b4-4be3-aa80-429fe1212566.png)



Pipeline steps
![Screenshot 2022-02-18 at 11 56 47](https://user-images.githubusercontent.com/18519371/154669892-237fae61-3991-433d-8107-abc0c49a6c17.png)

- diagram source - https://pastebin.com/raw/5KhePmK0 (just copy and paste in tldraw)
### Important design choices
* The CLI has no positional arguments and instead it is using the concept of required options
* The outputs (as shown in the diagram) will be put in the following folder `$output_dir/$source/$date/`
   * `output_dir` is specified via the CLI (`--output-dir`)
   * `source` is also specified via the CLI
   * `date` is a concatenation of the below two dates (using the `_` separator for joining) (e.g. `2022-02_2022-02-15`):
     * `from-month` (format `%Y-%m`) (specified via the CLI)
     * today (format `%Y-%m-%d`)
* The `PerformFilteringTask` creates symlinks to those articles that made it through the filtering and puts these symlinks to `filtered/`. This makes it very easy for `ParseTask` to ingest them.
* The mesh database needs to be provided via the CLI via `--mesh-topic-db=mesh_topic_db.json` -> The `bbs_database parse-mesh-rdf` is not included in the pipeline.
* The code disables the central scheduler (via `local_scheduler=True`). It is mostly for debugging purposes. However, we might want to use the central scheduler in "production". See more info https://luigi.readthedocs.io/en/stable/central_scheduler.html 
* Luigi can be very verbose. Its logging level was hardcoded to `WARNING`. Feel free to modify it to have more information on what `luigi` is doing.
## How to test?
First of all, initialize an empty database.

```bash
bbs_database init my_db.sqlite
```
Make sure you define the topic rules in `rules_config.jsonl`. If you want all articles to be accepted (debugging) just do
```json
{"label": "accept"}
```

You can create a helper script `launch_luigi.sh`.
```bash
 bbs_database run \
     --source=pubmed \
     --from-month=2022-02 \
     --filter-config=rules_config.jsonl \
     --output-dir=pipeline_results/ \
     --mesh-topic-db=mesh_topic_db.json \
     --db-url=my_db.sqlite \
     --grobid-host=dgx1.bbp.epfl.ch \
     --grobid-port=8070 \
     $@
```
And you can then run `luigi` in 2 different ways
**Dry run**
```bash
./launch_luigi.sh -n
```
* Shows what steps are still to do (`PENDING`) and which are done (`COMPLETED`).

**Real run**
```bash
./launch_luigi.sh
```

To speed up the pipelines (debugging purposes) consider creating the following script`custom_timeout`.
```bash
timeout 15 $@ || ( [[ $? -eq 124 ]] && exit 0 )
```
You can then simply use it inside of `ExternalProgramTask` as the first argument. Very useful for `DownloadTask`. Make sure that the Python process can find it in the `PATH` environment variable.

## TODO + To discuss
- [x] Write end to end tests (e.g. make the raw download as real as possible). Think of a way how to run "test" runs without having to download everything (maybe `timeout` but need to handle the nonzero exit code somehow)
   -  The best solution was to timeout the `DownloadTask` so that there are not that many articles
- [x] Handle logging and verbosity (`luigi` vs our entrypoints)
   - For now we just automatically set the logging level of all entrypoints to `logging.INFO`. However, can be changed using the hardcoder "global" variable `VERBOSITY`
- [x] Make sure no unzipping needed for `pubmed` (there is a PR for this)
   - [x] No unzipping needed for `topic-extract`
   - [x] No unzipping needed for `parse`
- [x] Verify that `biorxiv` and `medrxiv` work end-to-end
- [x] Best way to provide optional parameters to our entrypoints (e.g. `luigi.cfg`)
   - Let's do this in a different PR
   - e.g. the number of processes for `convert-pdf`
- [x] Test on `--db-type=postgres`
- [x] `bbs_database run --dry-run` remove the parameters to make it more readable
   - The parameters are important. 
- [x] Should we replace `ExternalProgramTask` with simple `from bluesearch.entrypoint.database.? import run; run()`? Maybe we don't need to replace it at all.
- [x] Try to make `luigi` history work
   - Let's do this in a different PR if useful 
- [x] What happens if the pipeline does not run within 1 day?
  - Solved via a global variable that stores a unique folder for the entire process 
- [x] Add `luigi` to requirements
- [ ] Test with more realistic `rules.json` 
- [ ] Make it possible to stop early

Please provide here instructions on how to test the changes introduced by this PR.
(if some changes cannot be tested by automated tests)

## Checklist

  (if it is not the case, please create an issue first).
  (if needed)
  (if needed)
  (if needed)
  (if a function is added or modified)
